stream_alert/shared/rule.py

```diff
@@ -0,0 +1,187 @@
+"""
+Copyright 2017-present, Airbnb Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from stream_alert.shared import LOGGER
+from stream_alert.shared.stats import time_rule
+
+
+class RuleInvalid(Exception):
+    """Exeception to raise for any errors with invalid rules"""
+
+
+def rule(**opts):
+    """Decorator to be used to register a rule"""
+    def decorator(rule_func):
+        """Rule decorator logic that returns instance of Rule"""
+        return Rule(rule_func, **opts)
+    return decorator
+
+
+def disable(rule_instance):
+    """Decorator to be used for disabling a rule"""
+    Rule.disable(rule_instance.rule_name)
+    return rule_instance
+
+
+class Rule(object):
+    """Rule class to handle processing"""
+    _rules = {}
+
+    def __init__(self, func, **kwargs):
+        self.func = func
+        self.rule_name = func.__name__
+        self.logs = kwargs.get('logs')
+        self.outputs = kwargs.get('outputs')
+        self.matchers = kwargs.get('matchers')
+        self.datatypes = kwargs.get('datatypes')
+        self.req_subkeys = kwargs.get('req_subkeys')
+        self.initial_context = kwargs.get('context')
+        self.context = None
+        self.disabled = False
+
+        if not (self.logs or self.datatypes):
+            raise RuleInvalid(
```

Comment:
To be consistent with `AlertCreationError`, should we rename this to `RuleCreationError`? Or `InvalidRuleError`? Either way, I like custom errors ending with `Error`, just like the built-in ones (`TypeError`, `ValueError`), etc.

Reply:
I like it! will make the change